src/main.c

```diff
@@ -505,18 +519,73 @@ static const struct xdg_toplevel_listener xdg_toplevel_listener = {
 	.close = xdg_toplevel_close,
 };
 
+static void wl_surface_enter(void* data, struct wl_surface* wl_surface,
+		struct wl_output* wl_output)
+{
+	struct window * w = data;
+	struct output* out;
+
+	wl_list_for_each(out, &outputs, link) {
+		if (out->wl_output == wl_output) {
+			wl_list_insert(&w->outputs, &out->link2);
+			return;
+		}
+	}
+
+	assert(false);
+}
+
+static void wl_surface_leave(void* data, struct wl_surface* wl_surface,
+		struct wl_output* wl_output)
+{
+	struct window * w = data;
+
+	struct output* out;
+	struct output* tmp;
+
+	wl_list_for_each_safe(out, tmp, &w->outputs, link2) {
```

Comment:
Maybe it's better to just mark the outputs instead using a variable named something like `has_surface`.